spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -152,22 +150,28 @@ protected void doStart() {
 
 		Object messageListener = containerProperties.getMessageListener();
 		Assert.state(messageListener != null, "A MessageListener is required");
-		if (messageListener instanceof GenericAcknowledgingMessageListener) {
-			this.acknowledgingMessageListener = (GenericAcknowledgingMessageListener<?>) messageListener;
-		}
-		else if (messageListener instanceof GenericMessageListener) {
-			this.listener = (GenericMessageListener<?>) messageListener;
-		}
-		else {
-			throw new IllegalStateException("messageListener must be 'MessageListener' "
-					+ "or 'AcknowledgingMessageListener', not " + messageListener.getClass().getName());
-		}
 		if (containerProperties.getConsumerTaskExecutor() == null) {
 			SimpleAsyncTaskExecutor consumerExecutor = new SimpleAsyncTaskExecutor(
 					(getBeanName() == null ? "" : getBeanName()) + "-C-");
 			containerProperties.setConsumerTaskExecutor(consumerExecutor);
 		}
-		this.listenerConsumer = new ListenerConsumer(this.listener, this.acknowledgingMessageListener);
+		Assert.state(messageListener instanceof GenericMessageListener, "Listener must be a GenericListener");
+		this.listener = (GenericMessageListener<?>) messageListener;
+		ListenerType listenerType = ListenerUtils.determineListenerType(this.listener);
+		if (this.listener instanceof DelegatingMessageListener) {
+			Object delegating = this.listener;
+			while (delegating instanceof DelegatingMessageListener) {
+				Object delegate = ((DelegatingMessageListener<?>) delegating).getDelegate();
+				listenerType = ListenerUtils.determineListenerType(delegate);
+				if (delegate instanceof DelegatingMessageListener) {
```

Comment:
No; we can't unconditionally cast the delegate's delegate to `DelegatingMessageListener` we have to test its type first.